### PR TITLE
Added param to toggle thruster visualizer. + thruster plugin cleanup

### LINKED ISSUE #26: Thruster Strength Artifacts
When the control system is killed but the simulator is still running some artifacts from the thruster strength bars are left after the sub floats back to the surface:
bug

### PR DESCRIPTION
**Pull request by dcrossman**
fixes #26

### CODE REVIEW

**Review by skallaher — APPROVED** [review]
Looks good to me.

**Comment by dcrossman** [comment]
Hold up! Before this is merged I'm just gonna fix #26 as well.

**Review by skallaher — CHANGES_REQUESTED** [review]
Please fix #26.

File: src/thruster.cpp
Comment (bkallaher) [inline src/thruster.cpp]: Maybe a better place for this is where updateVisualizers() is called on line 241.

**Review by bkallaher — CHANGES_REQUESTED** [review]

File: src/thruster.cpp
Comment (skallaher) [reply src/thruster.cpp]: Agreed. This would make it a little more readable.

**Comment by dcrossman** [comment]
I'm just gonna clean thrusters up in general.

**Comment by skallaher** [comment]
@BigDCross, be sure to rebase

File: src/thruster.cpp
Comment (skallaher) [inline src/thruster.cpp]: This is only loaded here on initialization, and so cannot be changed while the simulator is running. Also, because the value defaults to true and since the true default is loaded from the param file, this cannot be altered without changing the param file. Is this intended behavior?

**Comment by dcrossman** [comment]
Ok this took me longer than expected. In the course of fixing #26 I ran into an issue where the visualizers were disappearing if a visual message was not sent for some time, despite gazebo receiving a correct visual message (go to Window->Topic Visualization in gazebo to see gazebo messages). I think this is because gazebo was auto-disabling the visualizer entity but there's nothing in the visual message to fix this. So, in the end, I'm just publishing a visual message regardless of whether thruster visualization is on or not and just setting the visibility as appropriate.

**Comment by dcrossman** [comment]
Also the visualizers are actually red now!

**Comment by skallaher** [comment]
Just a note that in PalouseRobosub/robosub#118, I modified the vision launch file to set the visualize_thrusters param to false automatically if the argument `simulated` is true.

**Comment by bkallaher** [comment]
The visualizers don't move as smoothly as they used to any reason for that?

**Comment by skallaher** [comment]
Could it have to do with how often the visualization message is being published?

**Comment by bkallaher** [comment]
That was it. It seems like 10hz works a lot better than 5hz. But 10hz is still a little jumpy moving at high speeds.

File: src/thruster.cpp
Comment (skallaher) [inline src/thruster.cpp]: Perhaps make this a parameter? Or at least a constant somewhere. Either way, please update as per @bkallaher's comment.

**Comment by dcrossman** [comment]
Ok visualizer rate is a param now. Still don't merge this yet. Sometimes (at least on mine) the lines still disappear. And I still can't figure out why...

File: src/thruster.cpp
Comment (skallaher) [inline src/thruster.cpp:283]: Why are the parameters only being reloaded when the visualizers should update?

File: src/thruster.cpp
Comment (skallaher) [inline src/thruster.cpp:116]: You're inverting the `visualizer_update_rate here`, so your 0.1 becomes 10. Based upon [this](http://wiki.ros.org/roscpp/Overview/Time) documentation, that means your update time is every 10 seconds. @BigDCross, could this be a problem?

Comment (skallaher) [reply src/thruster.cpp:116]: I made a change of removing the `1.0/` and ran the simulator with the sub spinning and maintaining depth, I set the update rate to about .1 or .01 and saw no lagging of the visualization or any artifacts.

File: src/thruster.cpp
Comment (dcrossman) [reply src/thruster.cpp:283]: Because then they are only reloaded, like, 10 times a second rather than every frame. It's just the best place for the call.

File: src/thruster.cpp
Comment (dcrossman) [reply src/thruster.cpp:116]: I've got visualizer_update_rate set to 10 in the param file (as in update 10 times/second). Isn't that rate or is my terminology backward?

Comment (skallaher) [reply src/thruster.cpp:116]: That works, I would just change your hard coded default value.

File: src/thruster.cpp
Comment (skallaher) [reply src/thruster.cpp:283]: Unless you end up with your rate being once every 10 seconds, then it takes a while to reload.

File: src/thruster.cpp
Comment (dcrossman) [reply src/thruster.cpp:116]: If there's no default value the duration will end up being 0 and the loop will be run every frame (which slows gazebo to a near-halt)

File: src/thruster.cpp
Comment (dcrossman) [reply src/thruster.cpp:283]: Yea well...don't do that...

File: src/thruster.cpp
Comment (skallaher) [reply src/thruster.cpp:116]: You initialize `visualizer_update_rate` on line 110 to 0.1 and only warn about it if you can't find a parameter.

Comment (bkallaher) [reply src/thruster.cpp:116]: That value should probably be defaulted to 10 on line 110

Comment (dcrossman) [reply src/thruster.cpp:116]: Oh yea gotcha.

**Comment by skallaher** [comment]
Any ideas on what might be going wrong still?

**Comment by dcrossman** [comment]
The only thing I can think of is old gzserver instances that refuse to die eating all the messages. But otherwise I have no idea. Try it out yourself and see what happens. Maybe it's just me.

**Comment by skallaher** [comment]
What are the circumstances causing the bug? (i.e. What do I need to do to recreate it?)

**Comment by dcrossman** [comment]
Here is one way:
1. Start up simulator
2. Start control system
3. Send control msg (best to do this through rqt)
4. Kill and restart control system quickly (possibly do this multiple times)
I just did this and had 2/8 visualizers remain on screen. If you look at the visual messages under topic visualization in gazebo it seems like all the messages are being received and are correct.

So this happens even on a first run through of gazebo so it's not (or not just) old instances getting messages. 

Sometimes, though, the lines never appear at all.

**Comment by skallaher** [comment]
I followed your steps and haven't seen any visualizers become artifacts. I am on this branch for the simulator and dev for the robosub repo.

**Review by bkallaher — APPROVED** [review]

**Comment by bkallaher** [comment]
I got it to the point where the lines won't show up as well but I don't think it is a problem that we can solve.

**Review by skallaher — APPROVED** [review]

**Pull request merged**
